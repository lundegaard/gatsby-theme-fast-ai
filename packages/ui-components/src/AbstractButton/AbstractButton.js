import React, { forwardRef, useCallback } from 'react';
import PropTypes from 'prop-types';

const isNativeButton = el =>
	el.nodeName === 'BUTTON' || (el.nodeName === 'A' && el.getAttribute('href'));

/**
 * Keyword navigation for non native elements:
 * - Enter will trigger click event
 *
 * Adds aria attributes:
 * - `aria-disabled`
 */
export const useAbstractButton = ({
	disabled,
	onClick,
	onKeyDown,
	tabIndex = 0,
	...rest
}) => {
	const handleKeyDown = useCallback(
		event => {
			onKeyDown?.(event);

			if (
				event.target === event.currentTarget &&
				!disabled &&
				!isNativeButton(event.target) &&
				event.key === 'Enter'
			) {
				event.preventDefault();

				onClick?.(event);
			}
		},
		[onClick, disabled, onKeyDown],
	);

	return {
		onKeyDown: handleKeyDown,
		onClick,
		disabled,
		tabIndex: disabled ? -1 : tabIndex,
		'aria-disabled': disabled ? disabled : void 0,
		...rest,
	};
};

const AbstractButton = forwardRef(
	({ component: ButtonComponent = 'button', ...rest }, ref) => {
		const props = useAbstractButton({ ref, ...rest });

		return <ButtonComponent {...props} />;
	},
);

AbstractButton.propTypes = {
	component: PropTypes.elementType,
};

AbstractButton.displayName = 'AbstractButton';

export default AbstractButton;
