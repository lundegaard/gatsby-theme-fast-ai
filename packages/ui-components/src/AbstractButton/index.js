export { default } from './AbstractButton';
export { useAbstractButton } from './AbstractButton';
