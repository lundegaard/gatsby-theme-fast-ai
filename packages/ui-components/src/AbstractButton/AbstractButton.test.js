import React from 'react';
import { shallow } from 'enzyme';

import AbstractButton from './AbstractButton';

describe('<AbstractButton />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<AbstractButton {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
