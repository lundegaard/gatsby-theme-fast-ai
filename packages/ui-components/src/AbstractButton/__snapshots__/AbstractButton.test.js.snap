// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AbstractButton /> matches snapshot when shallowly rendered 1`] = `
<button
  onKeyDown={[Function]}
  tabIndex={0}
/>
`;
